pyro/contrib/conjugate/infer.py

```diff
@@ -0,0 +1,167 @@
+from functools import reduce
+from operator import mul
+
+import pyro.distributions as dist
+from pyro.distributions.util import sum_leftmost
+from pyro.poutine.messenger import Messenger
+from pyro.poutine.replay_messenger import ReplayMessenger
+
+
+def _make_cls(base, static_attrs, instance_attrs, parent_linkage=None):
+    r"""
+    Dynamically create classes named `_ + base.__name__`, which extend the
+    base class with other optional instance and class attributes, and have
+    a custom `.expand` method to propagate these attributes on expanded
+    instances.
+
+    :param cls base: Base class.
+    :param dict static_attrs: static attributes to add to class.
+    :param dict instance_attrs: instance attributes for initialization.
+    :param str parent_linkage: attribute in the parent class that holds
+        a reference to the distribution class.
+    :return cls: dynamically generated class.
+    """
+    def _expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(cls, _instance)
+        for attr in instance_attrs:
+            setattr(new, attr, getattr(self, attr))
+        if parent_linkage:
+            setattr(new.parent, parent_linkage, new)
+        return base.expand(self, batch_shape, _instance=new)
+
+    name = "_" + base.__name__
+    cls = type(name, (base,), instance_attrs)
+    for k, v in static_attrs.items():
+        setattr(cls, k, v)
+    cls.expand = _expand
+    return cls
+
+
+def _latent(base, parent):
+    return _make_cls(base, {"collapsible": True}, {"site_name": None, "parent": parent}, "_latent")
+
+
+def _conditional(base, parent):
+    return _make_cls(base, {"marginalize_latent": True}, {"parent": parent}, "_conditional")
+
+
+def _compound(base, parent):
+    return _make_cls(base, {}, {"parent": parent})
+
+
+class BetaBinomialPair(object):
+    def __init__(self):
+        self._latent = None
+        self._conditional = None
+
+    def latent(self, *args, **kwargs):
+        self._latent = _latent(dist.Beta, parent=self)(*args, **kwargs)
+        return self._latent
+
+    def conditional(self, *args, **kwargs):
+        self._conditional = _conditional(dist.Binomial, parent=self)(*args, **kwargs)
+        return self._conditional
+
+    def posterior(self, obs):
+        concentration1 = self._latent.concentration1
+        concentration0 = self._latent.concentration0
+        total_count = self._conditional.total_count
+        reduce_dims = len(obs.size()) - len(concentration1.size())
+        num_obs = reduce(mul, obs.size()[:reduce_dims], 1)
+        total_count = total_count[tuple(slice(1) if i < reduce_dims else slice(None)
+                                  for i in range(total_count.dim()))].reshape(concentration0.shape)
+        summed_obs = sum_leftmost(obs, reduce_dims)
+        return dist.Beta(concentration1 + summed_obs,
+                         num_obs * total_count + concentration0 - summed_obs,
```

Comment:
this is right if total_count is constant but I am not sure if it is still true when we have batch involved (because I don't understand the computation of total count in a few lines above.

Reply:
Good point. I think the requirement is that `total_count` shouldn't vary beyond the first k dims counting from the right, where `k = concentration0.dim()`. I have changed the `total_count` in the test below to illustrate this. If the `total_count` varies beyond the k dims, then this will throw a ValueError.

Reply:
According to [wikipedia](https://en.wikipedia.org/wiki/Conjugate_prior) posterior formula for BetaBinomial, I think that we need to `sum_leftmost` of `total_count` until it matches `concentration0.shape`. I don't understand why we do unexpand then multiply with num_obs here. Could you please elaborate more on it?

Reply:
I think that it would be enough to define
```
total_count = sum_leftmost(total_count, len(total_count.size()) - len(concentration1.size()))
```

Reply:
> I think that we need to sum_leftmost of total_count until it matches concentration0.shape. I don't understand why we do unexpand then multiply with num_obs here. Could you please elaborate more on it?

I think the issue here is that `total_count` will have the same shape as the observed data which might have a larger shape than `concentration0`, so we need to reduce it down appropriately. Suppose we have observations of shape `(5, 2) + concentration0.shape`. Suppose `total_count` is constant = 10. Then in actuality for each element in `concentration0`, we `10 (total_count) * 5 * 2  (left batch size) = 100 (num_obs)` observations, and the number of successes is given by `summed_obs =  sum_leftmost(obs, reduce_dims)`. That should be the same as the wikipedia formula? Let me know if that is still confusing.

Reply:
If we do unexpand for a constant tensor total_count=10, then it can not be reshaped into `concentration0.shape`. On the other hand, if `total_count.shape = obs.shape = probs.shape + 1`, then we can't do inference (while mathematically possible) because of the requirement that `total_count` shouldn't vary beyond the first k dims counting from the right.

Reply:
Note that total_count would be expanded internally by the distribution so it should be possible to reshape it into `concentration0.shape` if the leftmost dims come from expansion, but you are right that it is unnecessary and can be more general. Sorry, I missed your previous comment. I'm not even sure why I even had that in the previous PR. :) Just summing up the left dims should suffice!